kernel/solo5.h

```diff
@@ -31,69 +31,96 @@
- * issues (type confusion, missing const-ness, in/out parameters, inconsistent
- * return values). The comments in this file document the current
- * implementation.
+ * The value returned from this function is the application's exit status.
+ *
+ * FIXME: These APIs are to be considered unstable.  The comments in this file
+ * document the current implementation.
  */
-int solo5_app_main(char *cmdline);
+int solo5_app_main(const char *cmdline);
 
 /*
  * Network I/O.
+ *
+ * Currently only a single network device is supported. This device mimics an
+ * Ethernet v2 network interface with an MTU of 1500 bytes. Therefore, buffers
+ * passed to solo5_net_read_sync() must be at least 1514 bytes in length
+ * (including space for the Ethernet data link layer).
  */
 
 /*
- * Send a single network packet (*data) of size (n). Returns -1 on failure, 0
- * on success.
+ * Transmit a single network packet of (size) bytes from (*buf). (size) must be
+ * less than or equal to 1514 bytes.
+ *
+ * On failure:
+ *     Returns -1. This includes transient error conditions (e.g. short write)
+ *     and invalid arguments (e.g. size too big).
+ * On success:
+ *     Returns 0.
  */
-int solo5_net_write_sync(uint8_t *data, int n);
+int solo5_net_write_sync(const uint8_t *buf, size_t size);
 
 /*
- * Receive a single network packet into (*data) up to a maximum size of (*n).
- * Returns 0 on success and size of received packet (*n), -1 on failure or if
- * no packets are available.
+ * Receive a single network packet into (*buf), up to a maximum of (buf_size)
+ * bytes. (buf_size) must be at least 1514 bytes.
+ *
+ * On failure:
+ *     Returns -1. This includes invalid argument errors (e.g. buf_size too
+ *     small).
+ * If no packets are available:
+ *     Returns 0 and a (*read_size) of 0.
+ * On success:
+ *     Returns 0 and the length of the received packet in bytes in (*read_size).
  */
-int solo5_net_read_sync(uint8_t *data, int *n);
+int solo5_net_read_sync(uint8_t *restrict buf, size_t buf_size, size_t
+        *restrict read_size);
 
 /*
  * Returns a pointer to the network MAC address, formatted as a C string
  * XX:XX:XX:XX:XX:XX.
  */
-char *solo5_net_mac_str(void);
+const char *solo5_net_mac_str(void);
 
 /*
  * Block I/O.
+ *
+ * Currently only a single sector-addressable block device is supported, and
+ * only single-sector reads and writes are supported.
  */
 
 /*
- * Writes (n) bytes from (*data) starting at the sector (sec) to the block
- * device. (n) does not need to be a multiple of the sector size. Returns 0 on
- * success, -1 on error.
+ * Writes (size) bytes from (*buf) at the sector (offset) to the block device.
+ * (size) must be equal to the sector size.
+ *
+ * On failure:
+ *     Returns -1. This includes invalid argument errors and incomplete writes.
+ * On success:
+ *     Returns 0.
  */
-int solo5_blk_write_sync(uint64_t sec, uint8_t *data, int n);
+int solo5_blk_write_sync(uint64_t offset, const uint8_t *buf, size_t size);
```

Comment:
@mato yes that's right-- both of those cases will fail at runtime (possibly killing the app, not sure about that). We ought to define that behaviour a bit better! So perhaps it's more accurate to say that a runtime error exists (unfortunately) but there is one fewer case that can trigger it :) Maybe (big maybe) it would be possible to design the rest of errors out with clever use of types without being too annoying.